components/logins/src/db.rs

```diff
@@ -1768,4 +1832,54 @@ mod tests {
             }
         }
     }
+
+    #[test]
+    fn test_open_with_salt_create_db() {
+        let dir = tempdir::TempDir::new("open_with_salt").unwrap();
+        let dbpath = dir.path().join("logins.sqlite");
+        let dbpath = dbpath.to_str().unwrap();
+        let conn =
+            LoginDb::open_with_salt(dbpath, "testing", "952b9e3d53b39a8eba70b398acefa0a0").unwrap();
+        conn.query_one::<i64>("PRAGMA user_version").unwrap();
+    }
+
+    #[test]
+    fn test_get_salt_for_key() {
+        // First we create a database.
+        let dir = tempdir::TempDir::new("salt_for_key_test").unwrap();
+        let dbpath = dir.path().join("logins.sqlite");
+        let dbpath = dbpath.to_str().unwrap();
+        let conn = LoginDb::open(dbpath, Some("testing")).unwrap();
+        // Database created.
+        let expected_salt = conn.query_one::<String>("PRAGMA cipher_salt").unwrap();
+
+        let salt = LoginDb::open_and_get_salt(dbpath, "testing").unwrap();
+        assert_eq!(expected_salt, salt);
+    }
+
+    #[test]
+    fn test_get_salt_for_key_no_db() {
+        assert!(LoginDb::open_and_get_salt("nodbpath", "testing").is_err());
+    }
+
+    #[test]
+    fn test_plaintext_header_migration_full() {
+        // First we create a database.
+        let dir = tempdir::TempDir::new("plaintext_header_migration").unwrap();
+        let dbpath = dir.path().join("logins.sqlite");
+        let dbpath = dbpath.to_str().unwrap();
+        let conn = LoginDb::open(dbpath, Some("testing")).unwrap();
+        drop(conn);
+        // Database created.
+
+        // Step 1: get the salt.
+        let salt = LoginDb::open_and_get_salt(dbpath, "testing").unwrap();
+
+        // Step 2: migrate the db.
+        LoginDb::open_and_migrate_to_plaintext_header(dbpath, "testing", &salt).unwrap();
+
+        // Step 3: open using the salt.
+        let conn = LoginDb::open_with_salt(dbpath, "testing", &salt).unwrap();
+        conn.query_one::<i64>("PRAGMA user_version").unwrap();
+    }
```

Comment:
I'd like to see a couple more tests here:

* If the application passes in an inappropriate salt value, we should throw (this is especially important if we expect the app to generate its own salt for new databases).
* If we try to open a database with the wrong salt, it should throw.